File: drivers/node/singamdb.js
// singamdb - Official Node.js Native Driver for SingamDB Wire Protocol
const net = require('net');

const MAGIC = 0x534e474d; // 'SNGM'

const MessageType = {
  Ping: 1,
  Pong: 2,
  Handshake: 3,
  Insert: 10,
  Find: 11,
  GetById: 12,
  Update: 13,
  Delete: 14,
  CreateIndex: 20,
  Aggregate: 30,
  Explain: 40,
  ResponseOk: 100,
  ResponseError: 101
};

// Fast CRC32 table
const crcTable = new Uint32Array(256);
for (let i = 0; i < 256; i++) {
  let c = i;
  for (let k = 0; k < 8; k++) {
    c = ((c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1));
  }
  crcTable[i] = c;
}

function crc32(buf) {
  let crc = 0 ^ (-1);
  for (let i = 0; i < buf.length; i++) {
    crc = (crc >>> 8) ^ crcTable[(crc ^ buf[i]) & 0xFF];
  }
  return (crc ^ (-1)) >>> 0;
}

class SingamClient {
  constructor(url = 'singam://localhost:7778') {
    const parsed = new URL(url.replace('singam://', 'http://'));
    this.host = parsed.hostname || 'localhost';
    this.port = parseInt(parsed.port || '7778', 10);
    this.socket = null;
    this.reqIdCounter = 1;
    this.pendingRequests = new Map();
    this.buffer = Buffer.alloc(0);
    this.connected = false;
  }

  async connect() {
    return new Promise((resolve, reject) => {
      this.socket = net.createConnection({ host: this.host, port: this.port }, () => {
        this.connected = true;
        this.socket.setNoDelay(true);
        resolve(this);
      });

      this.socket.on('data', (chunk) => this._onData(chunk));
      this.socket.on('error', (err) => reject(err));
      this.socket.on('close', () => { this.connected = false; });
    });
  }

  _onData(chunk) {
    this.buffer = Buffer.concat([this.buffer, chunk]);

    while (this.buffer.length >= 16) {
      const magic = this.buffer.readUInt32BE(0);
      if (magic !== MAGIC) {
        throw new Error('Invalid SingamDB wire frame header magic.');
      }

      const msgType = this.buffer.readInt32BE(4);
      const reqId = this.buffer.readUInt32BE(8);
      const payloadLen = this.buffer.readInt32BE(12);

      const totalFrameLen = 16 + payloadLen + 4;
      if (this.buffer.length < totalFrameLen) {
        break; // Wait for full frame
      }

      const payload = this.buffer.slice(16, 16 + payloadLen);
      const receivedCrc = this.buffer.readUInt32BE(16 + payloadLen);

      // Verify CRC32
      const frameWithoutCrc = this.buffer.slice(0, 16 + payloadLen);
      const computedCrc = crc32(frameWithoutCrc);

      this.buffer = this.buffer.slice(totalFrameLen);

      const handler = this.pendingRequests.get(reqId);
      if (handler) {
        this.pendingRequests.delete(reqId);
        if (computedCrc !== receivedCrc) {
          handler.reject(new Error('CRC32 checksum mismatch in server response.'));
          continue;
        }

        const jsonStr = payload.toString('utf8');
        try {
          const parsed = JSON.parse(jsonStr);
          if (msgType === MessageType.ResponseError) {
            handler.reject(new Error(parsed.error || 'Server error'));
          } else {
            handler.resolve(parsed);
          }
        } catch (e) {
          handler.resolve(jsonStr);
        }
      }
    }
  }

  async _send(msgType, payloadObj) {
    if (!this.connected) await this.connect();

    const reqId = this.reqIdCounter++;
    const payloadStr = JSON.stringify(payloadObj);
    const payloadBytes = Buffer.from(payloadStr, 'utf8');

    const totalLen = 16 + payloadBytes.length + 4;
    const frame = Buffer.alloc(totalLen);

    frame.writeUInt32BE(MAGIC, 0);
    frame.writeInt32BE(msgType, 4);
    frame.writeUInt32BE(reqId, 8);
    frame.writeInt32BE(payloadBytes.length, 12);
    payloadBytes.copy(frame, 16);

    const crc = crc32(frame.slice(0, totalLen - 4));
    frame.writeUInt32BE(crc, totalLen - 4);

    return new Promise((resolve, reject) => {
      this.pendingRequests.set(reqId, { resolve, reject });
      this.socket.write(frame);
    });
  }

  database(dbName) {
    return new SingamDatabase(this, dbName);
  }

  close() {
    if (this.socket) {
      this.socket.end();
      this.connected = false;
    }
  }
}

class SingamDatabase {
  constructor(client, name) {
    this.client = client;
    this.name = name;
  }

  collection(collName) {
    return new SingamCollection(this.client, this.name, collName);
  }
}

class SingamCollection {
  constructor(client, dbName, collName) {
    this.client = client;
    this.dbName = dbName;
    this.collName = collName;
  }

  async insertOne(document, customId = null) {
    return this.client._send(MessageType.Insert, {
      Database: this.dbName,
      Collection: this.collName,
      Document: document,
      CustomId: customId
    });
  }

  async insertMany(documents) {
    const results = [];
    for (const doc of documents) {
      results.push(await this.insertOne(doc));
    }
    return results;
  }

  find(filter = {}) {
    return new SingamCursor(this.client, this.dbName, this.collName, filter);
  }

  async findOne(filter = {}) {
    const results = await this.find(filter).limit(1).toArray();
    return results.length > 0 ? results[0] : null;
  }

  async getById(docId) {
    return this.client._send(MessageType.GetById, {
      Database: this.dbName,
      Collection: this.collName,
      DocId: docId
    });
  }

  async updateOne(docId, updateData, merge = true) {
    return this.client._send(MessageType.Update, {
      Database: this.dbName,
      Collection: this.collName,
      DocId: docId,
      UpdateData: updateData,
      Merge: merge
    });
  }

  async deleteOne(docId) {
    return this.client._send(MessageType.Delete, {
      Database: this.dbName,
      Collection: this.collName,
      DocId: docId
    });
  }

  async createIndex(field, options = {}) {
    return this.client._send(MessageType.CreateIndex, {
      Database: this.dbName,
      Collection: this.collName,
      Field: Array.isArray(field) ? null : field,
      IsBTree: !!options.isBTree,
      IsUnique: !!options.unique,
      IsComposite: Array.isArray(field),
      Fields: Array.isArray(field) ? field : null
    });
  }

  async aggregate(pipelineRequest, filter = null) {
    return this.client._send(MessageType.Aggregate, {
      Database: this.dbName,
      Collection: this.collName,
      Request: pipelineRequest,
      Filter: filter
    });
  }

  async explain(filter = {}) {
    return this.client._send(MessageType.Explain, {
      Database: this.dbName,
      Collection: this.collName,
      Filter: filter
    });
  }
}

class SingamCursor {
  constructor(client, dbName, collName, filter) {
    this.client = client;
    this.dbName = dbName;
    this.collName = collName;
    this.filter = filter;
    this._sortField = null;
    this._ascending = true;
    this._projectFields = null;
    this._limit = 100;
    this._skip = 0;
  }

  sort(fieldOrObj, order = 1) {
    if (typeof fieldOrObj === 'object') {
      const keys = Object.keys(fieldOrObj);
      if (keys.length > 0) {
        this._sortField = keys[0];
        this._ascending = fieldOrObj[keys[0]] >= 0;
      }
    } else {
      this._sortField = fieldOrObj;
      this._ascending = order >= 0;
    }
    return this;
  }

  project(...fields) {
    this._projectFields = fields.flat();
    return this;
  }

  limit(num) {
    this._limit = num;
    return this;
  }

  skip(num) {
    this._skip = num;
    return this;
  }

  async toArray() {
    return this.client._send(MessageType.Find, {
      Database: this.dbName,
      Collection: this.collName,
      Filter: this.filter,
      SortField: this._sortField,
      Ascending: this._ascending,
      ProjectFields: this._projectFields,
      Limit: this._limit,
      Skip: this._skip
    });
  }
}

module.exports = {
  SingamClient,
  SingamDatabase,
  SingamCollection,
  MessageType
};

File: drivers/python/singamdb.py
# singamdb - Official Python Native Driver for SingamDB Wire Protocol
import socket
import struct
import json
import zlib

MAGIC = 0x534E474D  # 'SNGM'

class MessageType:
    PING = 1
    PONG = 2
    HANDSHAKE = 3
    INSERT = 10
    FIND = 11
    GET_BY_ID = 12
    UPDATE = 13
    DELETE = 14
    CREATE_INDEX = 20
    AGGREGATE = 30
    EXPLAIN = 40
    RESPONSE_OK = 100
    RESPONSE_ERROR = 101

class SingamClient:
    def __init__(self, host="localhost", port=7778):
        if "://" in str(host):
            clean = host.replace("singam://", "")
            parts = clean.split(":")
            self.host = parts[0]
            self.port = int(parts[1]) if len(parts) > 1 else 7778
        else:
            self.host = host
            self.port = port
        self.sock = None
        self.req_id_counter = 1

    def connect(self):
        if self.sock is None:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            self.sock.connect((self.host, self.port))
        return self

    def _send(self, msg_type, payload_dict):
        self.connect()
        req_id = self.req_id_counter
        self.req_id_counter += 1

        payload_bytes = json.dumps(payload_dict).encode("utf-8")
        payload_len = len(payload_bytes)

        # Header: Magic(4), MsgType(4), ReqId(4), PayloadLen(4)
        header = struct.pack(">IIII", MAGIC, msg_type, req_id, payload_len)
        frame_without_crc = header + payload_bytes
        crc = zlib.crc32(frame_without_crc) & 0xFFFFFFFF
        full_frame = frame_without_crc + struct.pack(">I", crc)

        self.sock.sendall(full_frame)

        # Read Response Header
        resp_header = self._recv_exact(16)
        r_magic, r_type, r_req, r_len = struct.unpack(">IIII", resp_header)
        if r_magic != MAGIC:
            raise ValueError("Invalid SingamDB wire frame header magic.")

        resp_payload_and_crc = self._recv_exact(r_len + 4)
        resp_payload = resp_payload_and_crc[:r_len]
        received_crc = struct.unpack(">I", resp_payload_and_crc[r_len:])[0]

        # Verify CRC
        computed_crc = zlib.crc32(resp_header + resp_payload) & 0xFFFFFFFF
        if computed_crc != received_crc:
            raise ValueError("CRC32 checksum mismatch in server response.")

        resp_str = resp_payload.decode("utf-8")
        try:
            parsed = json.loads(resp_str)
            if r_type == MessageType.RESPONSE_ERROR:
                raise RuntimeError(parsed.get("error", "Database server error"))
            return parsed
        except json.JSONDecodeError:
            return resp_str

    def _recv_exact(self, length):
        data = bytearray()
        while len(data) < length:
            chunk = self.sock.recv(length - len(data))
            if not chunk:
                raise ConnectionError("Server disconnected unexpectedly.")
            data.extend(chunk)
        return bytes(data)

    def database(self, db_name):
        return SingamDatabase(self, db_name)

    def __getitem__(self, db_name):
        return self.database(db_name)

    def close(self):
        if self.sock:
            self.sock.close()
            self.sock = None

class SingamDatabase:
    def __init__(self, client, name):
        self.client = client
        self.name = name

    def collection(self, coll_name):
        return SingamCollection(self.client, self.name, coll_name)

    def __getitem__(self, coll_name):
        return self.collection(coll_name)

class SingamCollection:
    def __init__(self, client, db_name, coll_name):
        self.client = client
        self.db_name = db_name
        self.coll_name = coll_name

    def insert_one(self, document, custom_id=None):
        return self.client._send(MessageType.INSERT, {
            "Database": self.db_name,
            "Collection": self.coll_name,
            "Document": document,
            "CustomId": custom_id
        })

    def find(self, filter_dict=None):
        return SingamCursor(self.client, self.db_name, self.coll_name, filter_dict or {})

    def find_one(self, filter_dict=None):
        res = self.find(filter_dict).limit(1).to_list()
        return res[0] if res else None

    def get_by_id(self, doc_id):
        return self.client._send(MessageType.GET_BY_ID, {
            "Database": self.db_name,
            "Collection": self.coll_name,
            "DocId": doc_id
        })

    def update_one(self, doc_id, update_dict, merge=True):
        return self.client._send(MessageType.UPDATE, {
            "Database": self.db_name,
            "Collection": self.coll_name,
            "DocId": doc_id,
            "UpdateData": update_dict,
            "Merge": merge
        })

    def delete_one(self, doc_id):
        return self.client._send(MessageType.DELETE, {
            "Database": self.db_name,
            "Collection": self.coll_name,
            "DocId": doc_id
        })

    def create_index(self, field, is_btree=False):
        is_composite = isinstance(field, list)
        return self.client._send(MessageType.CREATE_INDEX, {
            "Database": self.db_name,
            "Collection": self.coll_name,
            "Field": field if not is_composite else None,
            "IsBTree": is_btree,
            "IsComposite": is_composite,
            "Fields": field if is_composite else None
        })

    def aggregate(self, pipeline_dict, filter_dict=None):
        return self.client._send(MessageType.AGGREGATE, {
            "Database": self.db_name,
            "Collection": self.coll_name,
            "Request": pipeline_dict,
            "Filter": filter_dict
        })

class SingamCursor:
    def __init__(self, client, db_name, coll_name, filter_dict):
        self.client = client
        self.db_name = db_name
        self.coll_name = coll_name
        self.filter = filter_dict
        self._sort_field = None
        self._ascending = True
        self._project_fields = None
        self._limit = 100
        self._skip = 0

    def sort(self, field, order=1):
        self._sort_field = field
        self._ascending = order >= 0
        return self

    def project(self, *fields):
        self._project_fields = list(fields)
        return self

    def limit(self, count):
        self._limit = count
        return self

    def skip(self, count):
        self._skip = count
        return self

    def to_list(self):
        return self.client._send(MessageType.FIND, {
            "Database": self.db_name,
            "Collection": self.coll_name,
            "Filter": self.filter,
            "SortField": self._sort_field,
            "Ascending": self._ascending,
            "ProjectFields": self._project_fields,
            "Limit": self._limit,
            "Skip": self._skip
        })

    def __iter__(self):
        return iter(self.to_list())
